### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by gloriallluo**
- Rust wrapping of DPDK mempool, mbuf, ethdev, etc.
- UDP user-space implementation

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
+name: CI
+
+on: [push]
```
Comment (rogercloud) [inline .github/workflows/main.yml]: on pull request

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+env:
+  CI_RUST_TOOLCHAIN: nightly-2022-07-01
```
Comment (rogercloud) [inline .github/workflows/main.yml:6]: What nightly features are we using?

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  DPDK_VERSION: dpdk-stable-21.11.1
+
+jobs:
+  check:
+    name: Code style and Clippy check
```
Comment (rogercloud) [inline .github/workflows/main.yml]: Please split fmt and clippy.

File: Cargo.lock [hunk 4]
```diff
@@ -0,0 +1,604 @@
+# This file is automatically @generated by Cargo.
```
Comment (rogercloud) [inline Cargo.lock]: Usually we don't include Cargo.lock file in the repo.

File: Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+description = "A Rust async wrapper for DPDK"
+keywords = ["async", "dpdk"]
+categories = ["asynchronous", "dpdk", "network-programming"]
+repository = "https://github.com/datenlord/async-dpdk"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+bytes = "1.2.1"
+dpdk-sys = { git = "https://github.com/datenlord/dpdk-sys", rev = "febf1cc" }
```
Comment (rogercloud) [inline Cargo.toml:15]: Please file an issue in dpdk-sys to track formal crate release on crates.io.

File: examples/pktmbuf.rs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+//! An example using Mempool and Mbuf.
+
+use async_dpdk::{eal, lcore, mbuf::Mbuf};
+
+fn main() {
+    eal::Config::new().enter().unwrap();
```
Comment (rogercloud) [inline examples/pktmbuf.rs:7]: For that it's an example, please add some comments to explain what you're doing.

File: examples/udp.rs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+#[tokio::main]
+async fn main() {
+    eal::Config::new()
+        .device_probe(&["10.2.3.0", "10.2.3.1"])
+        .enter()
+        .unwrap();
+    net_dev::device_start().unwrap();
+    let srv = tokio::task::spawn(server());
+    client().await;
```
Comment (rogercloud) [inline examples/udp.rs:41]: How do you make sure that server is ready when client sends the request.

File: rust-toolchain [hunk 8]
```diff
@@ -0,0 +1 @@
+nightly-2022-07-01
```
Comment (rogercloud) [inline rust-toolchain:1]: Same as above. Why nightly.

File: src/agent.rs [hunk 9]
```diff
@@ -0,0 +1,483 @@
...
+    /// `rte_ip_frag_death_row` pointer.
+    dr: NonNull<rte_ip_frag_death_row>,
+}
+
+#[allow(unsafe_code)]
+#[allow(clippy::missing_docs_in_private_items)]
+impl IpFragmentTable {
+    fn new(socket_id: u32) -> Result<Self> {
+        const MS_PER_S: u64 = 1000;
+        let max_cycles = unsafe { (rte_get_tsc_hz() + MS_PER_S - 1) / MS_PER_S * MS_PER_S };
```
Comment (rogercloud) [inline src/agent.rs:74]: Please add unsafe comment as it's a convention. Won't repeat this comment in the following code.

File: src/agent.rs [hunk 10]
```diff
@@ -0,0 +1,483 @@
...
+#[allow(unsafe_code)]
+impl Drop for IpFragmentTable {
+    fn drop(&mut self) {
+        unsafe { rte_ip_frag_table_destroy(self.as_mut_ptr()) };
+    }
+}
+
+#[allow(unsafe_code)]
+#[allow(clippy::missing_docs_in_private_items)]
+impl IpFragDeathRow {
```
Comment (rogercloud) [inline src/agent.rs:100]: Do we need a drop function to release the resource manually?

Comment (rogercloud) [inline src/agent.rs:103]: Do we need a drop function to release the resource?

File: src/agent.rs [hunk 11]
```diff
@@ -0,0 +1,483 @@
...
+            #[allow(clippy::shadow_unrelated, clippy::unwrap_used)] // they're related
+            let m = new_m.unwrap();
+            #[allow(clippy::single_match)] // more protocols
+            match proto_id {
+                IP_NEXT_PROTO_UDP => handle_ipv4_udp(m),
+                _ => {}
+            }
+        }
+        RTE_ETHER_TYPE_IPV6 | RTE_ETHER_TYPE_ARP => {}
+        ether_type => eprintln!("Unsupported ether type {ether_type:x}"),
```
Comment (rogercloud) [inline src/agent.rs]: use log instead?

File: .github/workflows/main.yml (hunk 3 above)
Comment (gloriallluo) [reply .github/workflows/main.yml]: Done.

File: Cargo.lock (hunk 4 above)
Comment (gloriallluo) [reply Cargo.lock]: It's removed now.

File: src/agent.rs (hunk 10 above)
Comment (gloriallluo) [reply src/agent.rs:103]: Oops I forgot that.

File: examples/udp.rs (hunk 7 above)
Comment (gloriallluo) [reply examples/udp.rs:41]: Make the server sleep for a while?

File: examples/pktmbuf.rs (hunk 6 above)
Comment (gloriallluo) [reply examples/pktmbuf.rs:7]: Some comments are added to both examples.

**Pull request merged**